File: solutions/c/rational-numbers/1/rational_numbers.c
#include "rational_numbers.h"
#include <stdio.h>

rational_t add(rational_t r1, rational_t r2){
    rational_t ret;
    ret.numerator = r1.numerator * r2.denominator + r2.numerator * r1.denominator;
    ret.denominator = r1.denominator * r2.denominator;
    if (ret.numerator==0) ret.denominator = 1;
    return ret;
}
rational_t subtract(rational_t r1, rational_t r2){
    rational_t ret;
    ret.numerator = r1.numerator * r2.denominator - r2.numerator * r1.denominator;;
    ret.denominator = r1.denominator * r2.denominator;
    if (ret.numerator==0) ret.denominator = 1;
    return ret;
}
rational_t multiply(rational_t r1, rational_t r2){
    rational_t ret;
 
    if (abs(r1.numerator)==abs(r2.denominator)) {
        r1.numerator=r1.numerator/abs(r1.numerator);
        r2.denominator=r2.denominator/abs(r2.denominator);
    }
    if (abs(r2.numerator)==abs(r1.denominator)) {
        r2.numerator=r2.numerator/abs(r2.numerator);
        r1.denominator=r1.denominator/abs(r1.denominator);
    } 

    ret.numerator = r1.numerator * r2.numerator;
    ret.denominator = r1.denominator * r2.denominator;
    if (ret.numerator==0) ret.denominator = 1;
    return ret;
}
rational_t divide(rational_t r1, rational_t r2){
    rational_t ret;
    rational_t rep2;
    rep2.denominator = r2.numerator;
    rep2.numerator = r2.denominator;
    ret = multiply(r1,rep2);
    if (ret.denominator<0 && ret.numerator>0) {
        ret.denominator=-ret.denominator;
        ret.numerator=-ret.numerator;
    }
    if (ret.denominator<0 && ret.numerator<0) {
        ret.denominator=-ret.denominator;
        ret.numerator=-ret.numerator;
    }
    return ret;
}
rational_t absolute(rational_t r1){
    rational_t ret;
        ret.numerator = abs(r1.numerator);
        ret.denominator = abs(r1.denominator);
    return ret;
}
rational_t exp_rational(rational_t r1, int16_t n){
    rational_t ret;
        ret.numerator = pow(r1.numerator, n);
        ret.denominator = pow(r1.denominator, n);
    return ret;
}
float exp_real(int16_t n, rational_t r1){
    float ret;
    float e;
    e = r1.numerator;
    e = e/r1.denominator;
    ret = pow(n,e);  
    return ret;
}
rational_t reduce(rational_t r1){
    rational_t ret;
        int16_t g=gcd (abs(r1.numerator),abs(r1.denominator));
        ret.numerator = r1.numerator/g;
        ret.denominator = r1.denominator/g;
        if (ret.denominator < 0) {
            ret.denominator = -ret.denominator;
            ret.numerator = -ret.numerator;
        }
    return ret;
}
int16_t gcd(int16_t a, int16_t b) {
    if (a == 0) return b;
    if (b==0) return b;
    if (a==b) return a;
    if (a > b) return gcd(a-b, b);
    return gcd(a,b-a);
}

File: solutions/csharp/rational-numbers/1/RationalNumbers.cs
using System;
using System.Diagnostics;

public static class RealNumberExtension
{
    public static double Expreal(this int realNumber, RationalNumber r)
    {
        return Math.Pow(Math.Pow(realNumber, r.Numerator()),1/r.Denominator());
    }
}

public struct RationalNumber
{
    double n;
    double d;
    public RationalNumber(int numerator, int denominator)
    {
        n = numerator;
        d = denominator;
        //double g = gcd(n,d);
        //n = n/g;
        //d = n/g;
    }

    public double Numerator() => n;
    public double Denominator() => d;
    public static RationalNumber operator +(RationalNumber r1, RationalNumber r2)
    {
        RationalNumber t = new RationalNumber();
        t.n = (r1.n * r2.d + r2.n * r1.d);
        t.d = r1.d * r2.d;
        return t.Reduce();
    }

    public static RationalNumber operator -(RationalNumber r1, RationalNumber r2)
    {
        RationalNumber t = new RationalNumber();
        t.n = (r1.n * r2.d - r2.n * r1.d);
        t.d = r1.d * r2.d;
        return t.Reduce();
    }

    public static RationalNumber operator *(RationalNumber r1, RationalNumber r2)
    {
        RationalNumber t = new RationalNumber();
        t.n = r1.n * r2.n;
        t.d = r1.d * r2.d;
        return t.Reduce();
    }

    public static RationalNumber operator /(RationalNumber r1, RationalNumber r2)
    {
        RationalNumber t = new RationalNumber();
        t.n = r1.n * r2.d;
        t.d = r1.d * r2.n;
        return t.Reduce();
    }

    public RationalNumber Abs()
    {
        RationalNumber t = new RationalNumber();
        if (n < 0 && d <0)
        {
            t.n = -n;
            t.d = -d;
        }
        else if (n < 0 && d>0)
        {
            t.n = -n;
            t.d = d;
        }
        else if (n >0 && d <0)
        {
            t.n = n;
            t.d = -d;
        } else
        {
            t.n = n;
            t.d = d;
        }
        return t.Reduce();
    }

    public RationalNumber Reduce()
    {
        RationalNumber t = new RationalNumber();
        double g = gcd(n,d);
        t.n = n/g;
        t.d = d/g;
        if (t.d < 0) {
            t.n = -1 * t.n;
            t.d = -1 * t.d;
        }
        if (t.n == 0) t.d = 1;
        return t;
    }

    public RationalNumber Exprational(int power)
    {
        RationalNumber t = new RationalNumber();
        t.n = Math.Pow(n, power);
        t.d = Math.Pow(d, power);
        return t.Reduce();
    }

    public double Expreal(int baseNumber)
    {
        return Math.Pow(Math.Pow(baseNumber,n),1/d);
    }

    public double gcd(double n1, double n2) {
        if (n2 == 0) return n1;
        else return gcd(n2, n1 % n2);
    }
}